subprojects/ivy/src/main/java/org/gradle/api/publish/ivy/internal/publication/DefaultIvyPublication.java

```diff
@@ -256,192 +265,294 @@ public void setAlias(boolean alias) {
 
     @Override
     public void from(SoftwareComponent component) {
-        if (this.component != null) {
+        if (getComponent().isPresent()) {
             throw new InvalidUserDataException(String.format("Ivy publication '%s' cannot include multiple components", name));
         }
-        this.component = (SoftwareComponentInternal) component;
+        getComponent().set((SoftwareComponentInternal) component);
+        getComponent().finalizeValue();
         artifactsOverridden = false;
+
         updateModuleDescriptorArtifact();
     }
 
+    // TODO: This method should be removed in favor of lazily adding artifacts to the publication state.
+    // This is currently blocked by Signing eagerly realizing the publication artifacts.
     private void populateFromComponent() {
         if (populated) {
             return;
         }
         populated = true;
-        if (component == null) {
-            return;
+        if (!artifactsOverridden && parsedComponent.isPresent()) {
+            mainArtifacts.addAll(parsedComponent.get().getArtifacts());
+            configurations.addAll(parsedComponent.get().getConfigurations());
```

Comment:
This should be in another if statement which is not guarded by `!artifactsOverridden`